Intermittent build failure

I am by no definition a java developer, nor to I have any experience with gradle, but I noticed an intermittent build (test?) failure, it's pretty uncommon, but by simply looping `gradle clean build copyRuntimeLibs` I've been able to recreate it several times.

I'll quantify what I mean by "uncommon". I had it happen twice within about 5 builds, so I set out to forcibly recreate it; the next one didn't happen until after over an hour of continuous builds (averaging 12 seconds a piece), but it does always eventually happen.

(this is what I used, pretty simple really)
```Bash
while gradle clean build copyRuntimeLibs; do echo 'AGAIN!'; done
```

```
java.lang.AssertionError: Did not queue all records! expected:<5> but was:<3>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:555)
	at com.monetate.koupler.PipeKouplerTest.test(PipeKouplerTest.java:29)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:105)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:56)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:64)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:49)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:106)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:360)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

And that number at the top: `<5> but was:<3>`, the 3 is not constant, I've gotten it with 4 or 2 as well (I guess that might be obvious, but it might not be, I'm not too familiar with the project).

I'm on OSX, 10.11.3, but I've also recreated the same issue (identical stack trace) in Linux Mint, and with the java docker container (which I think is debian jessie?).

--- a/src/test/java/com/monetate/koupler/MockKinesisEventProducer.java
+++ b/src/test/java/com/monetate/koupler/MockKinesisEventProducer.java
@@ -1,5 +1,7 @@
 package com.monetate.koupler;
 
+import static org.junit.Assert.assertEquals;
+
 import java.util.concurrent.atomic.AtomicInteger;
 
 public class MockKinesisEventProducer extends KinesisEventProducer {
@@ -8,9 +10,20 @@ public class MockKinesisEventProducer extends KinesisEventProducer {
     }
 
     public AtomicInteger COUNT = new AtomicInteger();
-    
+
     @Override
     public void queueEvent(String event) {
         COUNT.getAndIncrement();
+    }
+
+    public void waitFor(int expectedCount) throws InterruptedException {
+        long totalWaitTime = 0;
+        while (expectedCount != COUNT.get()) {
+            Thread.sleep(10);
+            totalWaitTime += 10;
+            if (totalWaitTime > 5000) {
+                assertEquals("Did not queue all records!", expectedCount, COUNT.get());
+            }
+        }
     }
 }

--- a/src/test/java/com/monetate/koupler/PipeKouplerTest.java
+++ b/src/test/java/com/monetate/koupler/PipeKouplerTest.java
@@ -24,9 +24,8 @@ public class PipeKouplerTest {
         
         for (int i=0; i < TOTAL_LINES; i++){
             this.sendEvent(outputStream, i);
-        }        
-        Thread.sleep(500);
-        assertEquals("Did not queue all records!", TOTAL_LINES, mockProducer.COUNT.get());     
+        }
+        mockProducer.waitFor(TOTAL_LINES);
 	}
 	
     public void sendEvent(PipedOutputStream out, int x) throws IOException {
@@ -34,5 +33,5 @@ public class PipeKouplerTest {
         byte[] sendData = offer.getBytes();
         out.write(sendData);
     }
-
+    
 }

--- a/src/test/java/com/monetate/koupler/TcpKouplerTest.java
+++ b/src/test/java/com/monetate/koupler/TcpKouplerTest.java
@@ -1,7 +1,5 @@
 package com.monetate.koupler;
 
-import static org.junit.Assert.assertEquals;
-
 import java.io.IOException;
 import java.io.OutputStream;
 import java.net.Socket;
@@ -9,8 +7,6 @@ import java.net.Socket;
 import org.junit.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-
-import com.monetate.koupler.TcpKoupler;
 
 public class TcpKouplerTest {
     private static final Logger LOGGER = LoggerFactory.getLogger(TcpKouplerTest.class);
@@ -31,8 +27,7 @@ public class TcpKouplerTest {
         }
         
         LOGGER.debug("Hang tight -- waiting for TCP packets");
-        Thread.sleep(1000);
-        assertEquals("Did not receive all messages over TCP!", TOTAL_LINES, mockProducer.COUNT.get());
+        mockProducer.waitFor(TOTAL_LINES);
         socket.close();
     }
 

--- a/src/test/java/com/monetate/koupler/UdpKouplerServerTest.java
+++ b/src/test/java/com/monetate/koupler/UdpKouplerServerTest.java
@@ -1,7 +1,5 @@
 package com.monetate.koupler;
 
-import static org.junit.Assert.assertEquals;
-
 import java.io.IOException;
 import java.net.DatagramPacket;
 import java.net.DatagramSocket;
@@ -33,8 +31,7 @@ public class UdpKouplerServerTest {
         }
         long stop = System.currentTimeMillis();
         LOGGER.debug("Emitted [{}] packets in [{}] milliseconds.  Hang tight, testing for dropped packets in UDP server.", TOTAL_PACKETS, stop - start);
-        Thread.sleep(1000); // wait 1 seconds
-        assertEquals("Did not receive all UDP packets!", TOTAL_PACKETS, UdpKoupler.PACKETS.get());
+        mockProducer.waitFor(TOTAL_PACKETS);
     }
 
     public void sendEvent(InetAddress ipAddress, DatagramSocket socket, int x) throws IOException {

